### PR TITLE
Randomize node used to download data files

### LINKED ISSUE #2863: Randomize node used to download data files
### Problem

The mirror node always validates at least 1/3 signatures from all nodes, but then only downloads the signed data file from the first node sorted by ID. This usually results in always downloading the file from `0.0.3`. To be more fair and to better distribute rows for CitusDB we should randomize the download order for data files.

### Solution

For each of the stream types:
- Add a `node_id` column to each stream file table and backfill it with `node_account_id - 3`
- Drop `node_account_id`
- Randomize download order
  - Consider using `TreeMultimap.create(Ordering.natural(), Ordering.allEqual())`
  - Consider using a random iterator when walking over values

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by xin-hedera**
**Description**:

This PR improves downloader to download stream data files from a random node

- Replace `nodeAccountId` with `nodeId` in stream files
- Randomly pick a node to download stream data file
- Change stateproof REST API to look up node account id from `address_book_entry`

**Related issue(s)**:

Fixes #2863 

**Notes for reviewer**:

- Tested locally to verify the node downloader picks is random.
- Can ignore the changes to the record_file*.sql files in rest test folder, need to change column `node_account_id` to `node_id` as well as the value

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/downloader/Downloader.java [hunk 1]
```diff
@@ -325,6 +326,8 @@ private void verifySigsAndDownloadDataFiles(Multimap<StreamFilename, StreamFileS
                 throw new SignatureVerificationException(ex.getMessage() + ": " + statusMapMessage);
             }
 
+            // Shuffle the signatures to randomly pick a node to download the data file from
+            Collections.shuffle(signatures);
```
Comment (xin-hedera) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/downloader/Downloader.java:330]: `O(n)` complexity on `ArrayList` since it implements `RandomAccess`

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/BlockNumberMigration.java [hunk 2]
```diff
@@ -60,7 +60,7 @@ public String getDescription() {
 
     @Override
     protected MigrationVersion getMinimumVersion() {
-        return MigrationVersion.fromVersion("1.61.1");
+        return MigrationVersion.fromVersion("1.67.0");
```
Comment (xin-hedera) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/migration/BlockNumberMigration.java:63]: need to bump up the minimum version since

- the migration uses `RecordFileRepository`
- record_file table is different between version 1.61.1 and 1.67.0

This is the easiest fix and it shouldn't create any issue.

**Review by edwin-greene — DISMISSED** [review]
Looks good!

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/downloader/Downloader.java [hunk 3]
```diff
@@ -376,7 +379,7 @@ private void verifySigsAndDownloadDataFiles(Multimap<StreamFilename, StreamFileS
                             nodeId, sigFilename, e);
                 } catch (TransientProviderException e) {
                     log.warn("Error downloading data file from node {} corresponding to {}. Will retry another node",
-                            nodeId, sigFilename, e.getMessage());
+                            nodeId, sigFilename, e);
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/downloader/Downloader.java:382]: Will this print a stack trace at warn level? Maybe we should keep as before and just add `{}`?

File: hedera-mirror-common/src/main/java/com/hedera/mirror/common/domain/StreamFile.java [hunk 4]
```diff
@@ -74,9 +72,9 @@ default String getMetadataHash() {
 
     void setName(String name);
 
-    EntityId getNodeAccountId();
+    Long getNodeId();
```
[the remark this answers is not in the document]
Comment (xin-hedera) [reply hedera-mirror-common/src/main/java/com/hedera/mirror/common/domain/StreamFile.java:75]: perhaps worth keeping it so as to catch the issue of not setting node id?

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/downloader/Downloader.java [hunk 5]
```diff
@@ -73,10 +75,19 @@
...
     private static final String HASH_TYPE_FILE = "File";
     private static final String HASH_TYPE_METADATA = "Metadata";
     private static final String HASH_TYPE_RUNNING = "Running";
 
+    private static final Comparator<StreamFileSignature> STREAM_FILE_SIGNATURE_COMPARATOR = (left, right) -> {
+        if (Objects.equals(left, right)) {
+            return 0;
+        }
+
+        return Ordering.arbitrary().compare(left, right);
```
Comment (xin-hedera) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/downloader/Downloader.java:88]: The `arbitrary` ordering compares the objects by its identity

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/downloader/Downloader.java (hunk 3 above)
Comment (xin-hedera) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/downloader/Downloader.java]: change to add `{}` instead of printing a stack trace

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/ConsensusNodeWrapper.java [hunk 6]
```diff
@@ -22,14 +22,16 @@
...
 import lombok.Value;
 
 import com.hedera.mirror.common.domain.addressbook.AddressBookEntry;
 import com.hedera.mirror.common.domain.addressbook.NodeStake;
 import com.hedera.mirror.common.domain.entity.EntityId;
 
+@Builder
 @Value
-final class ConsensusNodeWrapper implements ConsensusNode {
+public final class ConsensusNodeWrapper implements ConsensusNode {
```
Comment (steven-sheehy) [inline hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/addressbook/ConsensusNodeWrapper.java:34]: Undo this change. There's a separate `ConsensusNodeStub` for use in tests.

File: hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/downloader/Downloader.java (hunk 5 above)
Comment (steven-sheehy) [reply hedera-mirror-importer/src/main/java/com/hedera/mirror/importer/downloader/Downloader.java:88]: Maybe add a comment about this and why it's necessary.

**Review by steven-sheehy — APPROVED** [review]
LGTM

**Review by edwin-greene — APPROVED** [review]
Looks good
